gobblin-service/src/main/java/org/apache/gobblin/service/modules/flowgraph/BaseFlowGraph.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.modules.flowgraph;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.gobblin.annotation.Alpha;
+
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+
+/**
+ * A thread-safe implementation of {@link FlowGraph}. The implementation maintains the following data structures:
+ *    <p>nodes - the set of {@link DataNode}s in the graph</p>
+ *    <p>nodesToEdges - the mapping from each {@link DataNode} to its outgoing {@link FlowEdge}s</p>
+ *    <p>dataNodeMap - the mapping from a node identifier to the {@link DataNode} instance</p>
+ *    <p>flowEdgeMap - the mapping from a edge label to the {@link FlowEdge} instance</p>
+ */
+@Alpha
+@Slf4j
+public class BaseFlowGraph implements FlowGraph {
+  @Getter
+  private Set<DataNode> nodes = new HashSet<>();
```

Comment:
Is `dataNodeMap` sufficient to do what the set does?

Reply:
Yes good point! Fixed it.